tools/libboard/board_dc/board_dc.c

```diff
@@ -0,0 +1,329 @@
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <glob.h>
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <net/ethernet.h>
+#include <opae/properties.h>
+#include <opae/utils.h>
+#include <opae/fpga.h>
+#include <sys/ioctl.h>
+#include <netinet/ether.h>
+
+#include "safe_string/safe_string.h"
+#include "board_dc.h"
+
+#define MACADDR_LEN 17
+// sysfs paths
+#define SYSFS_BMCFW_VER                     "spi-*/spi_master/spi*/spi*.*/bmcfw_flash_ctrl/bmcfw_version"
+#define SYSFS_MAX10_VER                     "spi-*/spi_master/spi*/spi*.*/max10_version"
+#define SYSFS_MACADDR_PATH                  "spi-*/spi_master/spi*/spi*.*/mac_address"
+#define SYSFS_MACCNT_PATH                   "spi-*/spi_master/spi*/spi*.*/mac_count"
+
+
+
+
+// Read BMC firmware version
+fpga_result read_bmcfw_version(fpga_token token, char *bmcfw_ver, size_t len)
+{
+	fpga_result res                = FPGA_OK;
+	fpga_result resval             = FPGA_OK;
+	uint32_t size                  = 0;
+	char buf[FPGA_VAR_BUF_LEN]     = { 0 };
+	fpga_object bmcfw_object;
+
+	if (bmcfw_ver == NULL) {
+		OPAE_ERR("Invalid Input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaTokenGetObject(token, SYSFS_BMCFW_VER, &bmcfw_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get token object");
+		return res;
+	}
+
+	res = fpgaObjectGetSize(bmcfw_object, &size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object size ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	// Return error if object size bigger then buffer size
+	if (size > FPGA_VAR_BUF_LEN) {
+		OPAE_ERR("object size bigger then buffer size");
+		resval = FPGA_EXCEPTION;
+		goto out_destroy;
+	}
+
+	res = fpgaObjectRead(bmcfw_object, (uint8_t *)buf, 0, size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	res = parse_fw_ver(buf, bmcfw_ver, len);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to parse version ");
+		resval = res;
+	}
+
+
+out_destroy:
+	res = fpgaDestroyObject(&bmcfw_object);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to destroy object");
+	}
+
+	return resval;
+}
+
+fpga_result parse_fw_ver(char *buf, char *fw_ver, size_t len)
+{
+	uint32_t var               = 0;
+	fpga_result res            = FPGA_OK;
+	int retval                 = 0;
+	char *endptr               = NULL;
+
+	if (buf == NULL || fw_ver == NULL) {
+		OPAE_ERR("Invalid Input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	/* BMC FW version format reading
+	NIOS II Firmware Build 0x0 32 RW[23:0] 24 hFFFFFF Build version of NIOS II Firmware
+	NIOS FW is up e.g. 1.0.1 for first release
+	[31:24] 8hFF Firmware Support Revision - ASCII code
+	0xFF is the default value without NIOS FW, will be changed after NIOS FW is up
+	*/
+
+	errno = 0;
+	var = strtoul(buf, &endptr, 16);
+	if (!endptr || *endptr == '\0') {
```

Comment:
The check should be
```c
if (!endptr || *endptr != '\0') {
```

Reply:
In other words, on success, endptr will be set to the null terminator of the input string.  If it fails, then endptr points to the character that failed to parse.